lib/controllers/reviews-controller.js

```diff
@@ -0,0 +1,343 @@
+
+    const uri = IssueishDetailItem.buildURI(host, owner, repo, number, homeRepository.getWorkingDirectoryPath());
+    return this.props.workspace.open(uri, {pending: true, searchAllPanes: true});
+  }
+
+  showSummaries = () => new Promise(resolve => this.setState({summarySectionOpen: true}, resolve));
+
+  hideSummaries = () => new Promise(resolve => this.setState({summarySectionOpen: false}, resolve));
+
+  showComments = () => new Promise(resolve => this.setState({commentSectionOpen: true}, resolve));
+
+  hideComments = () => new Promise(resolve => this.setState({commentSectionOpen: false}, resolve));
+
+  showThreadID = commentID => new Promise(resolve => this.setState(prev => {
+    prev.threadIDsOpen.add(commentID);
+    return {};
+  }, resolve));
+
+  hideThreadID = commentID => new Promise(resolve => this.setState(prev => {
+    prev.threadIDsOpen.delete(commentID);
+    return {};
+  }, resolve));
+
+  resolveThread = async thread => {
+    if (thread.viewerCanResolve) {
+      // optimistically hide the thread to avoid jankiness;
+      // if the operation fails, the onError callback will revert it.
+      this.hideThreadID(thread.id);
+      try {
+        await resolveReviewThreadMutation(this.props.relay.environment, {
+          threadID: thread.id,
+          viewerID: this.props.viewer.id,
+          viewerLogin: this.props.viewer.login,
+        });
+        addEvent('resolve-comment-thread', {package: 'github'});
+      } catch (err) {
+        this.showThreadID(thread.id);
+        this.props.reportMutationErrors('Unable to resolve the comment thread', err);
+      }
+    }
+  }
+
+  unresolveThread = async thread => {
+    if (thread.viewerCanUnresolve) {
+      try {
+        await unresolveReviewThreadMutation(this.props.relay.environment, {
+          threadID: thread.id,
+          viewerID: this.props.viewer.id,
+          viewerLogin: this.props.viewer.login,
+        });
+        addEvent('unresolve-comment-thread', {package: 'github'});
+      } catch (err) {
+        this.props.reportMutationErrors('Unable to unresolve the comment thread', err);
+      }
+    }
+  }
+
+  addSingleComment = async (commentBody, threadID, replyToID, path, position, callbacks = {}) => {
+    let pendingReviewID = null;
+    try {
+      this.setState({postingToThreadID: threadID});
+
+      const reviewResult = await addReviewMutation(this.props.relay.environment, {
+        pullRequestID: this.props.pullRequest.id,
+        viewerID: this.props.viewer.id,
+      });
+      const reviewID = reviewResult.addPullRequestReview.reviewEdge.node.id;
+      pendingReviewID = reviewID;
+
+      const commentPromise = addReviewCommentMutation(this.props.relay.environment, {
+        body: commentBody,
+        inReplyTo: replyToID,
+        reviewID,
+        threadID,
+        viewerID: this.props.viewer.id,
+        path,
+        position,
+      });
+      if (callbacks.didSubmitComment) {
+        callbacks.didSubmitComment();
```

Comment:
Ahhhh great catch :fishing_pole_and_fish:. We used to handle this with the `.catch()` clause in `ReviewsView::submitReply` but that won't fire now because we catch the exception.

I've added a `didFailComment` callback and I'm using it to restore the comment body now :+1: